core/integration/container_test.go

```diff
@@ -3599,3 +3599,95 @@ func TestContainerParallelMutation(t *testing.T) {
 	require.Equal(t, res.Container.A.EnvVariable, "BAR")
 	require.Empty(t, res.Container.B, "BAR")
 }
+
+func TestContainerBuildMergesWithParents(t *testing.T) {
+	t.Parallel()
+
+	c, ctx := connect(t)
+
+	// Create a builder container
+	builderCtr := c.Directory().WithNewFile(
+		"Dockerfile",
+		`
+FROM node:alpine
+
+ENV FOO=BAR
+LABEL "com.example.test-should-replace"="foo"
+RUN node --version
+`,
+	)
+
+	// Create a container with envs variables and labels
+	testCtr := c.Container().
+		WithEnvVariable("BOOL", "DOG").
+		WithEnvVariable("FOO", "BAZ").
+		WithLabel("com.example.test-should-exist", "test").
+		WithLabel("com.example.test-should-replace", "bar").
+		WithExposedPort(5000).
+		Build(builderCtr)
+
+	envShouldExist, err := testCtr.EnvVariable(ctx, "BOOL")
+	require.NoError(t, err)
+	require.Equal(t, "DOG", envShouldExist)
+
+	envShouldBeReplaced, err := testCtr.EnvVariable(ctx, "FOO")
+	require.NoError(t, err)
+	require.Equal(t, "BAR", envShouldBeReplaced)
+
+	labelShouldExist, err := testCtr.Label(ctx, "com.example.test-should-exist")
+	require.NoError(t, err)
+	require.Equal(t, "test", labelShouldExist)
+
+	labelShouldBeReplaced, err := testCtr.Label(ctx, "com.example.test-should-replace")
+	require.NoError(t, err)
+	require.Equal(t, "foo", labelShouldBeReplaced)
+
+	ports, err := testCtr.ExposedPorts(ctx)
+	require.NoError(t, err)
+
+	port, err := ports[0].Port(ctx)
+	require.NoError(t, err)
+	require.Equal(t, 5000, port)
```

Comment:
Maybe add an `EXPOSE 6000` and see if it's there. Also, you can test if it gets replaced correctly by setting `EXPOSE 5000/udp` and checking the protocol.

Reply:
See https://github.com/dagger/dagger/pull/5052#issuecomment-1546814114